<rdar://65206398> Add documentation for Packages in CI

federicobucchi opened the pull request:
Add documentation for Packages in CI

### Motivation:

Developers might be interested on adding building Swift Packages in Continuous Integration workflows

### Modifications:

- Added Documentation/ContinousIntegration.md
- Added link to ContinousIntegration.md in Documentation/README.md

### Result:

- New file Documentation/ContinousIntegration.md
- New link to  Documentation/ContinousIntegration.md in Documentation/README.md

federicobucchi commented:
smoke test please

federicobucchi commented:
@swift-ci please smoke test

abertelrud commented:
Thanks for adding this!

neonichu commented:
There's some existing stuff here: https://developer.apple.com/documentation/swift_packages/building_swift_packages_or_apps_that_use_them_in_continuous_integration_workflows -- maybe we should link to that as well?

tomerd commented:
@swift-ci please smoke test macOS Platform

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,18 @@
+
+## Provide Credentials
+
+To resolve package dependencies that require authentication, or private packages, you need to provide credentials to your CI setup.
```
Most folks run into issue with this aspect of the setup (when using private repos for example) so this is the key point and we need to elaborate further on this.

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,18 @@
+## Use the Expected Version of a Package Dependency
+
+To ensure the *CI* workflow’s reliability, make sure it uses the appropriate version of package dependencies.
+Update each package dependency in `Package.resolved` and commit it to your *Git* repository to ensure it’s always up-to-date on the *CI* environment to prevent the *CI* from building your project with unexpected versions of package dependencies.
```
I actually think that telling folks to always commit `Package.resolved` is a double edged sword, I think we need to re-write this paragraph to explain what `Package.resolved` is and what are the pros/cons of checking it -in

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,18 @@
+## Overview
+
+*Continuous integration* (*CI*) is the process of automating and streamlining the building, analyzing, testing, archiving, and publishing of your apps to ensure that they're always in a releasable state.
+Most projects that contain or depend on Swift packages don't require additional configuration. However, be sure to commit your project's Package.resolved file to your *Git* repository to ensure a reliable *CI* workflow that always uses the expected version of a package dependency.
```
this sentence is redundant given we have a section below about this

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,18 @@
+
+*Continuous integration* (*CI*) is the process of automating and streamlining the building, analyzing, testing, archiving, and publishing of your apps to ensure that they're always in a releasable state.
+Most projects that contain or depend on Swift packages don't require additional configuration. However, be sure to commit your project's Package.resolved file to your *Git* repository to ensure a reliable *CI* workflow that always uses the expected version of a package dependency.
+If your project depends on packages that require authentication you may need to perform additional configuration.
```
this sentence is redundant given we have a section below about this

tomerd reviewed (changes requested):
thanks for putting together, see comments inline

tomerd commented on Documentation/README.md:
```diff
@@ -17,6 +17,7 @@ We’ve designed the system to make it really easy to share packages on services
     * [Dependency Hell](#dependency-hell)
 * [Usage](Usage.md)
 * [PackageDescription API](PackageDescription.md)
+* [Continous Integration](ContinousIntegration.md)
```
👍

federicobucchi commented:
@swift-ci please smoke test

federicobucchi commented:
@swift-ci please smoke test Linux Platform

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,42 @@
+## Use the Expected Version of a Package Dependency
+
+To ensure the *CI* workflow’s reliability, make sure it uses the appropriate version of package dependencies.
+SwiftPM records the result of dependency resolution in `Package.resolved` (at the top-level of the package) and it's used when performing dependency resolution (rather than having SwiftPM searching the latest eligible version of each package).  Running `swift package update` updates all dependencies to the latest eligible versions and updates the `Package.resolved`.  You can commit `Package.resol ...
```
this reads mice better now, thank you!

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,42 @@
+To ensure the *CI* workflow’s reliability, make sure it uses the appropriate version of package dependencies.
+SwiftPM records the result of dependency resolution in `Package.resolved` (at the top-level of the package) and it's used when performing dependency resolution (rather than having SwiftPM searching the latest eligible version of each package).  Running `swift package update` updates all dependencies to the latest eligible versions and updates the `Package.resolved`.  You can commit `Package.resol ...
+
+## Provide Credentials
```
I am not sure about this one. I think we should explain here that SwiftPM relies on git for authentication, and as such the CI machines need to be configure so that git has access to the credentials it needs to access the repositories.

paraphrasing from https://developer.apple.com/documentation/swift_packages/building_swift_packages_or_apps_that_use_them_in_continuous_integration_workflows:

```
[To access private repositories] use SSH–based Git URLs for your packages and configure your SSH credentials. Set up your known_hosts file in the ~/.ssh directory of the user that runs your CI tasks. SwiftPM honors your SSH configuration — there’s no additional setup required.
If your SSH keys are password-protected, add them to the SSH agent before invoking SwiftPM by modifying the SSH configuration file.
````

We should also mention that CI services like Jenkins, Github Action, TravisCI, CircleCI, etc all provide ways to set up credentials and maybe link to their guides

Finally we need to mention in a new section that users can also use `xcodebuild` when building on macOS based CI hosts. 

cc @neonichu

neonichu replied:
Yep, I agree. I think we should only have a generic description of the process here and link off to somewhere else for specifics since the setup can be different depending on the CI solution.

federicobucchi commented:
@tomerd @neonichu I made the **Provide Credentials** section more generic and I added links to the CI services.
I have also added the section for `xcodebuild`.

Can you please check if the links to the guides for the services are the ones we want and if what I put in `xcodebuild` is still correct?

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,27 @@
+
+## Provide Credentials
+
+To resolve package dependencies that require authentication, or private packages, you need to provide credentials to your CI setup.
```
```suggestion
To resolve package dependencies that require authentication, such as private packages, you need to provide credentials to your CI setup.
```

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,27 @@
+## Provide Credentials
+
+To resolve package dependencies that require authentication, or private packages, you need to provide credentials to your CI setup.
+Use SSH–based Git URLs for your packages and configure your SSH credentials. Set up your known_hosts file in the ~/.ssh directory of the user that runs your CI tasks. SwiftPM honors your SSH configuration - there's no additional setup required.
```
```suggestion
SwiftPM honors the machine's SSH configuration - there's no additional setup required on SwiftPM side. For private package, use the SSH–based Git URLs and configure SSH credentials. You may also need to set up a known_hosts file in the ~/.ssh directory of the user that runs your CI tasks.  
```

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,27 @@
+
+To resolve package dependencies that require authentication, or private packages, you need to provide credentials to your CI setup.
+Use SSH–based Git URLs for your packages and configure your SSH credentials. Set up your known_hosts file in the ~/.ssh directory of the user that runs your CI tasks. SwiftPM honors your SSH configuration - there's no additional setup required.
+If your SSH keys are password-protected, add them to the SSH agent before invoking SwiftPM by modifying the SSH configuration file.
```
```suggestion
```

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,27 @@
+Use SSH–based Git URLs for your packages and configure your SSH credentials. Set up your known_hosts file in the ~/.ssh directory of the user that runs your CI tasks. SwiftPM honors your SSH configuration - there's no additional setup required.
+If your SSH keys are password-protected, add them to the SSH agent before invoking SwiftPM by modifying the SSH configuration file.
+
+CI services like [Jenkins](https://www.jenkins.io/doc/book/using/using-credentials), [Github Action](https://docs.github.com/en/free-pro-team@latest/actions/reference/authentication-in-a-workflow), [TravisCI](https://docs.travis-ci.com/user/private-dependencies), [CircleCI](https://circleci.com/docs/2.0/gh-bb-integration/#security), are providing ways to set up credentials.
```
```suggestion
CI services like [Jenkins](https://www.jenkins.io/doc/book/using/using-credentials), [Github Action](https://docs.github.com/en/free-pro-team@latest/actions/reference/authentication-in-a-workflow), [TravisCI](https://docs.travis-ci.com/user/private-dependencies), [CircleCI](https://circleci.com/docs/2.0/gh-bb-integration/#security) provide ways to set up SSH keys or other techniques to access private repositories. Since SwiftPM uses git to clone the repositories there's no additional setup required on SwiftPM side, and SwiftPM will honor the machine's SSH and git configuration.
```

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,27 @@
+When building on macOS based CI hosts you can use the command-line tool `xcodebuild`.
+`xcodebuild` uses Xcode's built-in Git tooling to connect to repositories.  In many cases, you don't need to make changes to how xcodebuild connects to them. However, some use cases require you use the configuration you set for your Mac's Git installation (Some examples: URL remapping, Proxy configurations, Advanced SSH configurations).
+
+For more information on using xcodebuild, visit [this link](https://developer.apple.com/library/archive/technotes/tn2339/_index.html).
```
```suggestion
For more information on using xcodebuild in continuous integration , see [this link](https://developer.apple.com/documentation/swift_packages/building_swift_packages_or_apps_that_use_them_in_continuous_integration_workflows).
```

tomerd commented on Documentation/ContinousIntegration.md:
```diff
@@ -0,0 +1,27 @@
+
+## Using xcodebuild
+When building on macOS based CI hosts you can use the command-line tool `xcodebuild`.
+`xcodebuild` uses Xcode's built-in Git tooling to connect to repositories.  In many cases, you don't need to make changes to how xcodebuild connects to them. However, some use cases require you use the configuration you set for your Mac's Git installation (Some examples: URL remapping, Proxy configurations, Advanced SSH configurations).
```
```suggestion
`xcodebuild` uses Xcode's built-in Git tooling to connect to repositories.  In many cases, you don't need to make changes to how xcodebuild connects to them. However, some use cases require you use the configuration you set for your Mac's Git installation (Some examples: URL remapping, Proxy configurations, Advanced SSH configurations). To have xcodebuild use your Mac’s Git installation and configuration, pass `-scmProvider system` to the xcodebuild command.
```

tomerd commented:
@federicobucchi  added some suggestion for final text

tomerd reviewed (approved):
lgtm, thanks for putting together <3

tomerd commented:
@swift-ci please smoke test
